server/services/recategorisationFilter.ts

```diff
@@ -0,0 +1,161 @@
+    [LOW_ROSH]: 'Low RoSH',
+    [LOW_RISK_OF_ESCAPE]: 'Low risk of escape',
+    [NO_CURRENT_TERRORISM_OFFENCES]: 'No current terrorism offences',
+    [NO_ROTL_RESTRICTIONS_OR_SUSPENSIONS]: 'No ROTL restrictions or suspensions',
+    [NO_ADJUDICATIONS_IN_THE_LAST_3_MONTHS]: 'No adjudications in the last 3 months',
+    [NOT_MARKED_AS_NOT_FOR_RELEASE]: "Not marked as 'Not for release'",
+    [STANDARD_OR_ENHANCED_INCENTIVE_LEVEL]: 'Standard or Enhanced incentive level',
+    [TIME_LEFT_TO_SERVE_BETWEEN_12_WEEKS_AND_3_YEARS]: 'Time left to serve is between 12 weeks and 3 years',
+  },
+}
+
+const dateIsNotBetween12WeeksAnd3Years = (dateString: string): boolean =>
+  moment(dateString).isAfter(moment().add(3, 'years')) || moment(dateString).isBefore(moment().add(12, 'weeks'))
+
+const getDateNMonthsAgo = (n: number) => {
+  return moment().subtract(n, 'month').format('YYYY-MM-DD')
+}
+
+const loadAdjudicationsData = async (
+  prisoners,
+  nomisClient,
+  agencyId: string
+): Promise<NomisAdjudicationHearingDto[]> => {
+  const prisonerNumbers = prisoners.map(prisoner => prisoner.offenderNo)
+  const dateThreeMonthsAgo = getDateNMonthsAgo(3)
+  const dateTwoMonthsAgo = getDateNMonthsAgo(2)
+  const dateOneMonthAgo = getDateNMonthsAgo(1)
+  const [adjudicationsThreeMonthsAgo, adjudicationsTwoMonthsAgo, adjudicationsLastMonth] = await Promise.all([
+    nomisClient.getOffenderAdjudications(prisonerNumbers, dateThreeMonthsAgo, dateTwoMonthsAgo, agencyId),
+    nomisClient.getOffenderAdjudications(prisonerNumbers, dateTwoMonthsAgo, dateOneMonthAgo, agencyId),
+    nomisClient.getOffenderAdjudications(prisonerNumbers, dateOneMonthAgo, moment().format('YYYY-MM-DD'), agencyId),
+  ])
+  return [...adjudicationsThreeMonthsAgo, ...adjudicationsTwoMonthsAgo, ...adjudicationsLastMonth]
+}
+
+export const filterListOfPrisoners = async (
+  filters: RecategorisationHomeFilters,
+  prisoners,
+  prisonerSearchData: Map<number, RecategorisationPrisonerSearchDto>,
+  nomisClient,
+  agencyId: string
+) => {
+  const allFilters = Object.values(filters)?.flat() || []
+  if (allFilters.length <= 0) {
+    return prisoners
+  }
+  let offenderNumbersWithAdjudications = []
+  if (allFilters.includes(NO_ADJUDICATIONS_IN_THE_LAST_3_MONTHS)) {
+    const adjudicationsData = await loadAdjudicationsData(prisoners, nomisClient, agencyId)
+    offenderNumbersWithAdjudications = adjudicationsData.map(adjudicationsDatum => adjudicationsDatum.offenderNo)
+  }
+  return prisoners.filter(prisoner => {
+    const currentPrisonerSearchData = prisonerSearchData.get(prisoner.bookingId)
+    const activeNonExpiredAlertCodes =
+      (currentPrisonerSearchData &&
+        currentPrisonerSearchData.alerts
+          ?.filter(alert => alert.active && !alert.expired)
+          .map(alert => alert.alertCode)) ||
+      []
+    const incentiveLevelCode = currentPrisonerSearchData?.currentIncentive?.level.code
+    for (let i = 0; i < allFilters.length; i += 1) {
```

Comment:
possibly slightly nicer to use `for (const filter of allFilters)`

Reply:
I get an eslint error if I use for of: iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them